### PR TITLE
Build add machine form.

### PR DESCRIPTION
**Pull request by Caleb-Ellis**
## Done
- Built "Add machine" form. Extra MACs and Save and add another buttons to be implemented in #794 and #795 

## QA
- Go to /r/machines/add and check that the add machine form shows up.
- Check that the Minimum kernel field is correctly set to the default if that has been set - otherwise it should be "No minimum kernel"
- Check that the MAC address field validates correctly
- Choose different power types from the select and check that each field is validated correctly
- Check that you can add a machine if all the required fields have been filled. For simplicity's sake you can choose "Manual" from the Power type dropdown because it has no extra power parameter fields.
- Check that you are redirected to /r/machines with a notification confirming the machine was added.
- Go back to the form and enter a MAC address you know is already used. Submit the form and check that a server error displays at the top of the form.

Fixes canonical-web-and-design/MAAS-squad#1729

### Default
screenshot_25

### Client-side validation
screenshot_26

### Server errors
screenshot_28

### CODE REVIEW

File: ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  resourcepool as resourcePoolSelectors,
+  zone as zoneSelectors
+} from "app/base/selectors";
+import { useAddMessage, useWindowTitle } from "app/base/hooks";
+import AddMachineFormFields from "../AddMachineFormFields";
+import FormCard from "app/base/components/FormCard";
+import FormikForm from "app/base/components/FormikForm";
+import FormCardButtons from "app/base/components/FormCardButtons";
+
+const trimPowerParameters = (powerType, powerParameters) => {
```
Comment (huwshimi) [inline ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js]: Might be useful to leave a comment here about why we need to do this.

File: ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    dispatch(generalActions.fetchHweKernels());
+    dispatch(generalActions.fetchPowerTypes());
+    dispatch(resourcePoolActions.fetch());
+    dispatch(zoneActions.fetch());
+  }, [dispatch]);
+
+  // When selected power type changes, update the validation schema with new
+  // power_parameters shape.
+  useEffect(() => {
+    if (powerType) {
```
Comment (huwshimi) [inline ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js]: Do we also need to unset the schema if this gets set back to null or can't that happen?

**Comment by huwshimi** [comment]
@Caleb-Ellis I had trouble QAing this (due to my local maas, not this branch). If no one gets to this before Monday I'll rebuild my maas and try again.

**Comment by lilyvidenova** [comment]
👍

File: ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js (hunk 1 above)
Comment (Caleb-Ellis) [reply ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js]: Good idea. Added it in now.

File: ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js (hunk 2 above)
Comment (Caleb-Ellis) [reply ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js]: It shouldn't be possible to unset the power type once one is selected, but it doesn't hurt to add that check in anyway.

Fixed.

**Comment by huwshimi** [comment]
@Caleb-Ellis how are you testing adding machines? Do you have a snap maas or `make start`? Did you have to do any additional setup to be able to add machines? I've been trying with a multipass/make-start maas and a snap maas running on hardware and it fails with a myriad of errors e.g. at the moment my make-start maas responds with `[Errno 2] No such file or directory: '/home/multipass/code/maas/.run/etc/maas/drivers.yaml'` when I submit the add machine form.

**Comment by Caleb-Ellis** [comment]
@huwshimi I've been using karura pretty much exclusively since the UI separation was complete. I have run into that issue with the `drivers.yaml` file using a `make start`'d MAAS before though, but that usually gets fixed with a fresh install. FTR I've never actually set up a snap MAAS. What happens when you copy [this pastebin](https://pastebin.canonical.com/p/dVgx63s7gz/) into `/home/multipass/code/maas/.run/etc/maas/drivers.yaml`? It's the `drivers.yaml` file from my local MAAS - no idea what it does though hahaha

**Comment by squidsoup** [comment]
QA looks good, thanks @Caleb-Ellis!

**Pull request merged**